tuplex/codegen/src/FunctionRegistry.cc

```diff
@@ -966,6 +966,301 @@ namespace tuplex {
+            auto x_ty = input_types[0];
+            auto y_ty = input_types[1];
+
+            if (args.size() == 2) {
+                // rel_tol and abs_tol not specified
+                rel_tol_val = _env.f64Const(1e-09);
+                abs_tol_val = _env.f64Const(0);
+            } else if (args.size() == 3) {
+                // assume that the third argument is rel_tol
+                rel_tol_val = args[2].val;
+                abs_tol_val = _env.f64Const(0);
+            } else {
+                assert(args.size() == 4);
+                // assume that the third argument is rel_tol and the fourth argument is abs_tol
+                // this means we don't handle the case where abs_tol is specified but rel_tol is not
+                rel_tol_val = args[2].val;
+                abs_tol_val = args[3].val;
+            }
+
+            auto rel_tol = _env.upCast(builder, rel_tol_val, _env.doubleType());
+            auto abs_tol = _env.upCast(builder, abs_tol_val, _env.doubleType());
+
+            // check x and y types - bools and ints can be optimized!
+            // TODO: error check both rel_tol > 0 and abs_tol >= 0
+            if (x_ty == python::Type::BOOLEAN && y_ty == python::Type::BOOLEAN) {
+                auto xor_xy = builder.CreateXor(x_val.val, y_val.val);
+                auto rel_cmp = builder.CreateFCmpOGE(rel_tol, _env.f64Const(1));
+                auto abs_cmp = builder.CreateFCmpOGE(abs_tol, _env.f64Const(1));
+                auto rel_or_abs = builder.CreateOr(rel_cmp, abs_cmp);
+                auto eq_check = builder.CreateXor(xor_xy, _env.boolConst(true));
+                auto or_res = builder.CreateOr(rel_or_abs, eq_check);
+
+                auto resVal = _env.upcastToBoolean(builder, or_res);
+                auto resSize = _env.i64Const(sizeof(int64_t));
+
+                return SerializableValue(resVal, resSize);
+            } else if (x_ty != python::Type::F64 && y_ty != python::Type::F64) {
+                // cast x/y to integers
+                auto x = _env.upCast(builder, x_val.val, _env.i32Type());
+                auto y = _env.upCast(builder, y_val.val, _env.i32Type());
+
+                auto cur_block = builder.GetInsertBlock();
+                assert(cur_block);
+
+                // create new blocks for each case
+                BasicBlock *bb_l1 = BasicBlock::Create(builder.getContext(), "opt_<_1", builder.GetInsertBlock()->getParent());
+                BasicBlock *bb_standard = BasicBlock::Create(builder.getContext(), "opt_standard", builder.GetInsertBlock()->getParent());
+                BasicBlock *bb_done = BasicBlock::Create(builder.getContext(), "cmp_done", builder.GetInsertBlock()->getParent());
+
+                // allocate space for return value
+                auto val = _env.CreateFirstBlockAlloca(builder, _env.i1Type());
+
+                // first block comparison (x ?== y)
+                auto eq_res = builder.CreateICmpEQ(x, y);
```

Comment:
why do we need floating point instructions in this case? shouldn't this be solvable using integer instructions solely?